File: backend/app/api/routes/scripts.py
from __future__ import annotations

from pathlib import Path

from fastapi import APIRouter, Depends, File, Form, UploadFile
from sqlalchemy import Engine
from sqlalchemy.orm import sessionmaker

from app.api.deps import get_runtime_root, get_session_factory
from app.core.response import success_response
from app.schemas.script_registry import ScriptListResponse, ScriptMainFileRequest, ScriptProfileRequest, ScriptResponse, ScriptRunConfigRequest, ScriptToggleRequest, ScriptUpdateRequest
from app.services.script_service import ScriptService


router = APIRouter(tags=["scripts"])


def build_script_service(
    session_factory: sessionmaker = Depends(get_session_factory),
    runtime_root: Path = Depends(get_runtime_root),
) -> ScriptService:
    engine: Engine = session_factory.kw["bind"]
    return ScriptService(engine=engine, runtime_root=runtime_root)


@router.get("/scripts")
def list_scripts(service: ScriptService = Depends(build_script_service)) -> dict[str, object]:
    data = ScriptListResponse(items=[ScriptResponse.model_validate(item) for item in service.list_scripts()])
    return success_response(data.model_dump())


@router.post("/scripts/upload")
async def upload_script(
    script_name: str = Form(...),
    script_code: str | None = Form(default=None),
    script_type: str = Form(...),
    platform: str = Form(...),
    version: str | None = Form(default=None),
    description: str | None = Form(default=None),
    profile_key: str | None = Form(default=None),
    script_file: UploadFile = File(...),
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    content = await script_file.read()
    result = service.upload(
        script_name=script_name,
        script_code=script_code,
        script_type=script_type,
        platform=platform,
        version=version,
        description=description,
        profile_key=profile_key,
        filename=script_file.filename or "script.py",
        content=content,
    )
    return success_response(ScriptResponse.model_validate(result).model_dump())


@router.post("/scripts/{script_code}/toggle")
def toggle_script(
    script_code: str,
    payload: ScriptToggleRequest,
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    result = service.set_enabled(script_code=script_code, enabled=payload.enabled)
    return success_response(ScriptResponse.model_validate(result).model_dump())


@router.patch("/scripts/{script_code}/profile")
def update_script_profile(
    script_code: str,
    payload: ScriptProfileRequest,
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    result = service.update_profile(script_code=script_code, profile_key=payload.profile_key)
    return success_response(ScriptResponse.model_validate(result).model_dump())


@router.patch("/scripts/{script_code}/run-config")
def update_script_run_config(
    script_code: str,
    payload: ScriptRunConfigRequest,
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    result = service.update_run_config(
        script_code=script_code,
        default_run_mode=payload.default_run_mode,
        default_cdp_port=payload.default_cdp_port,
        supports_pause=payload.supports_pause,
        supports_cancel=payload.supports_cancel,
        default_timeout_seconds=payload.default_timeout_seconds,
    )
    return success_response(ScriptResponse.model_validate(result).model_dump())


@router.get("/scripts/{script_code}/files")
def list_script_files(
    script_code: str,
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    files = service.list_script_files(script_code=script_code)
    return success_response(files)


@router.patch("/scripts/{script_code}/main-file")
def update_script_main_file(
    script_code: str,
    payload: ScriptMainFileRequest,
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    result = service.update_main_file(script_code=script_code, main_file=payload.main_file)
    return success_response(ScriptResponse.model_validate(result).model_dump())


@router.get("/scripts/cdp-port-status")
def get_cdp_port_status(
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    data = service.check_cdp_ports()
    return success_response(data)


@router.post("/scripts/{script_code}/clone")
def clone_script(
    script_code: str,
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    result = service.clone_script(script_code=script_code)
    return success_response(ScriptResponse.model_validate(result).model_dump())


@router.patch("/scripts/{script_code}/update")
def update_script_meta(
    script_code: str,
    payload: ScriptUpdateRequest,
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    result = service.update_script(
        script_code=script_code,
        script_name=payload.script_name,
        script_type=payload.script_type,
        platform=payload.platform,
        description=payload.description,
    )
    return success_response(ScriptResponse.model_validate(result).model_dump())


@router.delete("/scripts/{script_code}")
def delete_script(
    script_code: str,
    service: ScriptService = Depends(build_script_service),
) -> dict[str, object]:
    service.delete_script(script_code=script_code)
    return success_response(None)

File: frontend/src/api/scripts.ts
import { apiRequest } from "@/api/http";

export interface ScriptItem {
  id: number;
  script_code: string;
  script_name: string;
  script_type: string;
  platform: string;
  version: string | null;
  profile_key: string | null;
  script_dir: string;
  absolute_dir: string;
  main_file: string;
  enabled: boolean;
  default_run_mode: string | null;
  default_cdp_port: number | null;
  supports_pause: boolean;
  supports_cancel: boolean;
  default_timeout_seconds: number;
  description: string | null;
  updated_at: string | null;
}

export interface ScriptUploadPayload {
  scriptName: string;
  scriptType: string;
  platform: string;
  description: string;
  profileKey?: string | null;
  file: File;
}

export interface ScriptRunConfigPayload {
  default_run_mode?: string | null;
  default_cdp_port?: number | null;
  supports_pause?: boolean;
  supports_cancel?: boolean;
  default_timeout_seconds?: number;
}

interface ScriptListResponse {
  items: ScriptItem[];
}

export function fetchScripts(): Promise<ScriptListResponse> {
  return apiRequest<ScriptListResponse>("/scripts");
}

export function uploadScript(payload: ScriptUploadPayload): Promise<ScriptItem> {
  const formData = new FormData();
  formData.append("script_name", payload.scriptName);
  formData.append("script_code", "");
  formData.append("script_type", payload.scriptType);
  formData.append("platform", payload.platform);
  formData.append("version", "");
  formData.append("description", payload.description.trim());
  if (payload.profileKey) formData.append("profile_key", payload.profileKey);
  formData.append("script_file", payload.file);

  return apiRequest<ScriptItem>("/scripts/upload", {
    method: "POST",
    body: formData,
  });
}

export function updateScriptProfile(scriptCode: string, profileKey: string | null): Promise<ScriptItem> {
  return apiRequest<ScriptItem>(`/scripts/${scriptCode}/profile`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ profile_key: profileKey }),
  });
}

export function toggleScript(scriptCode: string, enabled: boolean): Promise<ScriptItem> {
  return apiRequest<ScriptItem>(`/scripts/${scriptCode}/toggle`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ enabled }),
  });
}

export function updateScriptRunConfig(scriptCode: string, payload: ScriptRunConfigPayload): Promise<ScriptItem> {
  return apiRequest<ScriptItem>(`/scripts/${scriptCode}/run-config`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

export function fetchScriptFiles(scriptCode: string): Promise<string[]> {
  return apiRequest<string[]>(`/scripts/${scriptCode}/files`);
}

export function updateScriptMainFile(scriptCode: string, mainFile: string): Promise<ScriptItem> {
  return apiRequest<ScriptItem>(`/scripts/${scriptCode}/main-file`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ main_file: mainFile }),
  });
}

export function deleteScript(scriptCode: string): Promise<void> {
  return apiRequest<void>(`/scripts/${scriptCode}`, { method: "DELETE" });
}

export interface CdpPortStatus {
  port: number;
  script_name: string;
  script_code: string;
  in_use: boolean;
}

export function fetchCdpPortStatus(): Promise<CdpPortStatus[]> {
  return apiRequest<CdpPortStatus[]>("/scripts/cdp-port-status");
}

export function cloneScript(scriptCode: string): Promise<ScriptItem> {
  return apiRequest<ScriptItem>(`/scripts/${scriptCode}/clone`, {
    method: "POST",
  });
}

export function updateScript(
  scriptCode: string,
  payload: { script_name?: string; script_type?: string; platform?: string; description?: string | null },
): Promise<ScriptItem> {
  return apiRequest<ScriptItem>(`/scripts/${scriptCode}/update`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}
